src/osd/ReplicatedBackend.cc

```diff
@@ -1713,13 +1717,16 @@ void ReplicatedBackend::submit_push_data(
     t->omap_setheader(target_coll, recovery_info.soid, omap_header);
   }
   uint64_t off = 0;
+  uint32_t fadvise_flags = CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL;
+  if (!get_parent()->object_has_io_pending_recovery(recovery_info.soid))
```

Comment:
This isn't valid on the replica.  waiting_for_unreadable and waiting_for_degraded are only maintained on the primary.  You need to restructure how this information is passed around.  I'd also very much prefer that the call into the backend indicate whether the recovery IO should be marked as dont_need or not.  I don't really like poking back up into the parent to check the waiting_for_* queues.  The parent has the information available when it calls into the backend.